src/rust/engine/serverset/src/lib.rs

```diff
@@ -220,65 +279,83 @@ impl<T: Clone + Send + Sync + 'static> Serverset<T> {
   pub fn next(&self) -> BoxFuture<(T, HealthReportToken), String> {
     let now = Instant::now();
     let mut inner = self.inner.lock();
-    let server_count = inner.connections.len();
+
+    let existing_connections = inner.connections.len();
+    if existing_connections < inner.connection_limit && inner.servers.len() > existing_connections {
+      // Find a healthy server without a connection, connect to it.
+      if let Some(ret) = inner.connect() {
+        inner.next_connection = inner.next_connection.wrapping_add(1);
+        return futures::future::ok(ret).to_boxed();
+      }
+    }
+
+    if !inner.connections.is_empty() {
+      let next_connection = inner.next_connection;
+      inner.next_connection = inner.next_connection.wrapping_add(1);
+      let connection = &inner.connections[next_connection % inner.connections.len()];
+      return futures::future::ok((
+        connection.connection.clone(),
+        HealthReportToken {
+          server_index: connection.server_index,
+        },
+      ))
+      .to_boxed();
+    }
 
     let mut earliest_future = None;
-    for _ in 0..server_count {
-      let i = inner.next % server_count;
-      inner.next = inner.next.wrapping_add(1);
-      let server = &inner.connections[i];
+    for _ in 0..inner.servers.len() {
+      let i = inner.next_server % inner.servers.len();
+      inner.next_server = inner.next_server.wrapping_add(1);
+      let server = &inner.servers[i];
       let unhealthy_info = &server.unhealthy_info;
       if let Some(ref unhealthy_info) = unhealthy_info {
-        // Server is unhealthy. Note when it will become healthy.
-
         let healthy_at = unhealthy_info.healthy_at();
 
         if healthy_at > now {
-          let healthy_sooner_than_previous = if let Some((_, previous_healthy_at)) = earliest_future
-          {
+          let healthy_sooner_than_previous = if let Some(previous_healthy_at) = earliest_future {
             previous_healthy_at > healthy_at
           } else {
             true
           };
 
           if healthy_sooner_than_previous {
-            earliest_future = Some((i, healthy_at));
+            earliest_future = Some(healthy_at);
           }
-          continue;
         }
-      }
-      // A healthy server! Use it!
-      return futures::future::ok((server.server.clone(), HealthReportToken { index: i }))
-        .to_boxed();
+      } // else cannot happen, or we would already have connected and returned.
     }
     // Unwrap is safe because if we hadn't populated earliest_future, we would already have returned.
```

Comment:
I get this is not new in this changeset, so don't feel like you have to act on it, but it feels like instead of having comments like this one where the ownace is on the reader to verify all of this logic, avoiding early returns and having `else` branches to match the `if` branches would make this an actual no-brainer.

Reply:
Rephrased this in a more iteratory way, but it still has an `unwrap` because the ordering of the checks we're doing makes it hard to encode the required information cleanly into the type system.

We want to say:
1. Maybe connect (iterate to find a connectable)
2. Else maybe reuse
3. Else iterate to find what will next be connectable

If we were happy to unconditionally do the next finding in the maybe-connect stage, even if we may end up reusing a connection, we could phrase this as:

1. Iterate over all servers, finding the next one we should try to connect to
2. If we can make a connection, and the server is healthy, connect to it
3. If nothing is connectable, and we have an open connection, re-use it
4. Else reschedule for when something will be connectable

But because we don't want to do the iteration and time calculations in step 1 if step 2 is going to fail and we're going to use step 3, we can't just write this as one cute expression.